File: src/csharp/parser/Parser.cs
namespace O7;

using System;
using System.Collections.Generic;
using System.Text;
using static ByteArrayUtils;
using ParseResult = System.Tuple<ASTUntyped, ParseError>;

public class PreParser {
    public static ParseResult parse(Expr inp) {
        var resultCurr = new ListStatements();
        var result = resultCurr;
        var resultBacktrack = new Stack<Tuple<ListStatements, int>>();
        var reservedWords = getReservedMap();
        var coreOperators = getOperatorList();
        var backtrack = new Stack<Tuple<ListExpr, int>>();

        if (inp is ListExpr le) {
            int i = 0;
            int j = 0;
            backtrack.push(new Tuple<ListExpr, int>(le, i));
            resultBacktrack.push(new Tuple<ListStatements, int>(resultCurr, i));
            var curr = le;
            while (backtrack.peek() != null) {
                var back = backtrack.pop();
                curr = back.Item1;
                i = back.Item2;

                var backResult = resultBacktrack.pop();
                resultCurr = backResult.Item1;
                j = backResult.Item2;

                while (i < curr.val.Count) {
                    l(i.ToString());
                    if (curr.val[i] is ListExpr le2) {
                        backtrack.push(new Tuple<ListExpr, int>(curr, i + 1));
                        resultBacktrack.push(new Tuple<ListStatements, int>(resultCurr, j + 1));

                        if (le2.pType == ExprLexicalType.curlyBraces) {
                            ListStatements newList = new ListStatements(SubexprType.list);
                            resultCurr.val.Add(newList);
                            newList.sType = SubexprType.list;

                            resultCurr = newList;
                        } else if (le2.pType == ExprLexicalType.dataInitializer) {
                            var newElem = new DataInitializer();
                            resultCurr.val.Add(newElem);

                            resultCurr = newElem;
                        } else { // if (le2.pType == ExprLexicalType.parens || le2.pType == ExprLexicalType.statement) {
                            ListStatements newStatement = new ListStatements(SubexprType.parens);
                            resultCurr.val.Add(newStatement);
                            result = newStatement;
                        }
                        curr = le2;
                        i = 0;
                        j = 0;
                    } else {
                        resultCurr.val.Add(parseAtom(curr.val[i], reservedWords, coreOperators));
                        ++i;
                        ++j;
                    }
                }

            }
        } else {
            return new Tuple<ASTUntyped, ParseError>(parseAtom(inp, reservedWords, coreOperators), null);
        }
        return new Tuple<ASTUntyped, ParseError>(result, null);

    }

    static Dictionary<String, ReservedType> getReservedMap() {
        var result = new Dictionary<String, ReservedType>();
        foreach (ReservedType enumValue in Enum.GetValues(typeof(ReservedType))) {
            result[enumValue.ToString()] = enumValue;
        }
        return result;
    }

    static List<Tuple<List<OperatorSymb>, CoreOperator>> getOperatorList() {
        var result = new List<Tuple<List<OperatorSymb>, CoreOperator>>();
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.plus },                             CoreOperator.plus));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.minus},                             CoreOperator.minus));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.asterisk},                          CoreOperator.times));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.slash},                             CoreOperator.divideBy));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.ampersand, OperatorSymb.ampersand}, CoreOperator.and));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.pipe, OperatorSymb.pipe},           CoreOperator.or));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.exclamation},                       CoreOperator.not));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.ampersand},                         CoreOperator.bitwiseAnd));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.pipe},                              CoreOperator.bitwiseOr));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.tilde, OperatorSymb.ampersand},     CoreOperator.bitwiseNot));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.caret},                             CoreOperator.bitwiseXor));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.equals},                            CoreOperator.defineImm));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.colon, OperatorSymb.equals},        CoreOperator.defineMut));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.lt, OperatorSymb.minus},            CoreOperator.assignmentMut));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.plus, OperatorSymb.equals},         CoreOperator.plusMut));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.minus, OperatorSymb.equals},        CoreOperator.minusMut));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.asterisk, OperatorSymb.equals},     CoreOperator.timesMut));
        result.Add(new Tuple<List<OperatorSymb>, CoreOperator>(new List<OperatorSymb>() { OperatorSymb.slash, OperatorSymb.equals},        CoreOperator.divideMut));
        return result;
    }


    static bool arraysEqual(List<OperatorSymb> a, List<OperatorSymb> b) {
        if (a.isEmpty() && a.Count != b.Count) return false;
        for (int i = 0; i < a.Count; ++i) {
            if (a[i] != b[i]) return false;
        }
        return true;
    }


    static ASTUntyped parseAtom(Expr inp,
                                Dictionary<String, ReservedType> reservedWords,
                                List<Tuple<List<OperatorSymb>, CoreOperator>> coreOperators) {
        if (inp is IntToken it) {
            return new IntLiteral(it.val);
        } else if (inp is FloatToken ft) {
            return new FloatLiteral(ft.val);
        } else if (inp is StringToken st) {
            return new StringLiteral(st.val);
        } else if (inp is WordToken wt) {
            var str = Encoding.ASCII.GetString(wt.val);
            if (str == "true") return new BoolLiteral(true);
            if (str == "false") return new BoolLiteral(false);
            if (reservedWords.TryGetValue(str, out ReservedType resWord)) {
                return new Reserved(resWord);
            } else {
                return new Ident(str);
            }
        } else if (inp is OperatorToken ot) {
            foreach (var oper in coreOperators) {
                if (arraysEqual(ot.val, oper.Item1)) {
                    return new CoreOperatorAST(oper.Item2);
                }
            }
            return new OperatorAST(ot.val);
        } else {
            // should never happen
            return new IntLiteral(-1);
        }
    }
}

File: src/main/java/tech/sozonov/o7/parser/Parser.java
package tech.sozonov.o7.parser;
import java.util.Map;
import java.util.Optional;
import tech.sozonov.o7.lexer.types.LexicalContext;
import tech.sozonov.o7.lexer.types.OperatorSymb;
import tech.sozonov.o7.lexer.types.Expr.*;
import tech.sozonov.o7.parser.types.Syntax;
import tech.sozonov.o7.parser.types.AST.*;
import static tech.sozonov.o7.parser.types.SyntaxContexts.SyntaxContext.*;
import tech.sozonov.o7.parser.types.SyntaxContexts.SyntaxContext;
import tech.sozonov.o7.parser.types.SyntaxError;
import tech.sozonov.o7.utils.Tuple;
import tech.sozonov.o7.utils.Either;
import tech.sozonov.o7.utils.Stack;
import static tech.sozonov.o7.utils.ArrayUtils.*;
import static tech.sozonov.o7.utils.ListUtils.*;
import lombok.val;


public class Parser {

/**
 * All the structure from the lexer stays, but new structure is added.
 * For example, what was "if x > 5 -> (f)" becomes "if (x > 5) (f)".
 * Full list of new punctuation symbols: -> : $
 */
public static Tuple<ASTBase, SyntaxError> parse(final ExprBase inp) {
    var result = new ASTList(curlyBraces);
    val syntax = new Syntax();
    var backtrack = new Stack<Tuple<ListExpr, Integer>>();
    var resultBacktrack = new Stack<ASTList>();

    if (!(inp instanceof ListExpr le)) {
        return new Tuple<ASTBase, SyntaxError>(parseAtom(inp, syntax), null);
    }

    int i = 0;
    backtrack.push(new Tuple<ListExpr, Integer>(le, i));
    resultBacktrack.push(result);

    var currInput = le;
    ASTList curr = result;

    while (backtrack.peek() != null) {
        val back = backtrack.pop();
        currInput = back.i0;
        i = back.i1;

        while (i < currInput.val.size()) {
            val сurrToken = currInput.val.get(i);
            if (сurrToken instanceof ListExpr le2) {
                backtrack.push(new Tuple<ListExpr, Integer>(currInput, i + 1));
                boolean skipFirstToken = false;

                // curlyBrace -> curlyBrace or nothing (if directly inside a bounded core form)
                // stmt -> funcall | assignment | core form
                // paren -> funcall | core form
                // dataInit -> dataInit

                if (le2.lType == LexicalContext.curlyBraces) {
                    if (curr.isUnbounded()) {
                        if (curr.isEmpty()) return new Tuple<>(result, new SyntaxError("Curly braces are not allowed in an unbounded core syntax form like " + curr.ctx));
                        curr = cleanPop(resultBacktrack, curr);
                    } else {
                        if (curr.isBounded()) curr.newStatement();
                        val newList = new ASTList(curlyBraces);
                        curr.add(newList);
                        resultBacktrack.push(curr);
                        curr = newList;
                    }
                } else if (le2.lType == LexicalContext.statement) {
                    if (curr.ctx != curlyBraces && !curr.isCoreForm()) {
                        return new Tuple<>(result, new SyntaxError("Statements are only allowed in curly braces or core syntax forms, not inside " + curr.ctx));
                    }
                    if (curr.isUnbounded()) {
                        if (curr.statementFitsCoreForm(le2)) {
                            curr.newStatement();
                        } else {
                            curr = cleanPop(resultBacktrack, curr);
                        }
                    }
                    if (curr.isBounded()) curr.newStatement();
                    if (!curr.isClauseBased()) {

                        val listType = determineListType(le2, syntax.contexts);
                        if (listType.isLeft()) return new Tuple<>(result, listType.getLeft());
                        skipFirstToken = listType.get().i1;

                        val newList = new ASTList(listType.get().i0);
                        val mbError = curr.add(newList);
                        if (mbError.isPresent()) return new Tuple<>(result, mbError.get());

                        resultBacktrack.push(curr);
                        curr = newList;
                    }
                } else if (le2.lType == LexicalContext.parens) {
                    if (curr.isUnbounded()) {
                        if (curr.isEmpty()) new Tuple<>(result, new SyntaxError("Parentheses are not allowed in an unbounded core syntax form like " + curr.ctx));
                    }
                    if (curr.isBounded()) curr.newStatement();
                    if (curr.ctx == curlyBraces) return new Tuple<>(result, new SyntaxError("Parentheses are not allowed directly in curly braces"));

                    val listType = determineListType(le2, syntax.contexts);
                    if (listType.isLeft()) return new Tuple<>(result, listType.getLeft());

                    skipFirstToken = listType.get().i1;

                    val newList = new ASTList(listType.get().i0);
                    val mbError = curr.add(newList);
                    if (mbError.isPresent()) return new Tuple<>(result, mbError.get());

                    resultBacktrack.push(curr);
                    curr = newList;
                } else {
                    if (curr.isBounded()) {
                        return new Tuple<>(result, new SyntaxError("Data initializers are not allowed in a bounded core syntax form like " + curr.ctx));
                    }

                    val dataInit = new ASTList(dataInitializer);
                    val mbError = curr.add(dataInit);
                    if (mbError.isPresent()) return new Tuple<>(result, mbError.get());

                    resultBacktrack.push(curr);
                    curr = dataInit;
                }

                currInput = le2;
                i = skipFirstToken ? 1 : 0;
            } else {
                val atom = parseAtom(сurrToken, syntax);
                val mbError = curr.addAtom(atom);
                if (mbError.isPresent()) return new Tuple<>(result, mbError.get());
                ++i;
            }
        }

        if (i == currInput.val.size()) {
            val mbSaturated = curr.gotSaturated();
            if (mbSaturated.isPresent()) {
                if (mbSaturated.get()) {
                    curr = cleanPop(resultBacktrack, curr);
                }
            } else {
                // should never happen
                return new Tuple<>(result, new SyntaxError("Strange error: oversaturated syntax form"));
            }
        }
    }
    if (resultBacktrack.peek() != null) {
        return new Tuple<>(result, new SyntaxError("Strange error: non-empty result stack at the end"));
    }
    return new Tuple<ASTBase, SyntaxError>(result, null);
}


static ASTList cleanPop(final Stack<ASTList> backtrack, final ASTList curr) {
    if (last(curr.data).isEmpty()) {
        removeLast(curr.data);
    }
    return backtrack.pop();
}

/**
 * Pre-condition: the input must be either a Statement or a Parens.
 * Determines the syntactic type of the expression: a function call, an assignment/definition, a core syntactic form, or a type declaration.
 * Returns: either a parse error,
 * or a tuple of SyntaxContext and a boolean of whether to skip the first token (necessary for the case of a core syntax form).
 */
static Tuple<SyntaxContext, Boolean> determineListTypeNoErrCheck(ListExpr input, Map<String, SyntaxContext> syntaxContexts) {
    val mbCore = checkListIsCore(input, syntaxContexts);
    if (mbCore.isPresent())  return new Tuple<>(mbCore.get(), true);

    if (input.val.size() >= 3 && input.val.get(1) instanceof OperatorToken ot) {
        val firstOper = ot.val.get(0);
        if (ot.val.size() == 1 && firstOper == OperatorSymb.equals) {
            return new Tuple<>(assignImmutable, false);
        } else if (ot.val.size() == 2 && ot.val.get(1) == OperatorSymb.equals) {
            if (firstOper == OperatorSymb.plus) return new Tuple<>(assignMutablePlus, false);
            if (firstOper == OperatorSymb.minus) return new Tuple<>(assignMutableMinus, false);
            if (firstOper == OperatorSymb.asterisk) return new Tuple<>(assignMutableTimes, false);
            if (firstOper == OperatorSymb.slash) return new Tuple<>(assignMutableDiv, false);
        }
        if (ot.val.size() == 2 && firstOper == OperatorSymb.colon && ot.val.get(1) == OperatorSymb.colon) return new Tuple<>(typeDeclaration, false);
    }
    if (input.val.size() == 1 && input.val.get(0) instanceof ListExpr le && le.lType == LexicalContext.curlyBraces) {
        return new Tuple<>(curlyBraces, false);
    }
    return new Tuple<>(funcall, false);
}

/**
 * Pre-condition: the input must be either a Statement or a Parens.
 * Determines the syntactic type of the expression: a function call, an assignment/definition, a core syntactic form, or a type declaration.
 * Returns: either a parse error, or a tuple of SyntaxContext and a boolean of whether to skip the first token (useful for core syntax forms).
 */
static Either<SyntaxError, Tuple<SyntaxContext, Boolean>> determineListType(ListExpr input, Map<String, SyntaxContext> syntaxContexts) {
    val tuple = determineListTypeNoErrCheck(input, syntaxContexts);
    if (input.lType == LexicalContext.parens && ASTBase.isAssignment(tuple.i0)) {
        return Either.left(new SyntaxError("Assignments are not allowed in parentheses, only on the statement level (i.e. inside curly braces)"));
    } else {
        return Either.right(tuple);
    }
}


/**
 * Determines core syntactical forms, but only judging on the first token (or first 2 tokens for unbounded-able forms).
 */
static Optional<SyntaxContext> checkListIsCore(ListExpr expr, Map<String, SyntaxContext> reservedWords) {
    if (!(expr.val.get(0) instanceof WordToken)) return Optional.empty();
    val word = (WordToken)expr.val.get(0);
    if (!reservedWords.containsKey(word.val)) return Optional.empty();

    val coreSyntaxContext = reservedWords.get(word.val);
    return Optional.of(coreSyntaxContext);
}


static ASTBase parseAtom(ExprBase inp, Syntax syntax) {
    if (inp instanceof IntToken it) {
        return new IntLiteral(it.val);
    } else if (inp instanceof FloatToken ft) {
        return new FloatLiteral(ft.val);
    } else if (inp instanceof StringToken st) {
        return new StringLiteral(st.val);
    } else if (inp instanceof WordToken wt) {
        var str = wt.val;
        if (str == "true") return new BoolLiteral(true);
        if (str == "false") return new BoolLiteral(false);

        val resWord = syntax.reservedWords.getOrDefault(str, null);
        if (resWord != null) return new ReservedLiteral(resWord);

        val coreForm = syntax.contexts.getOrDefault(str, null);
        if (coreForm != null) return new ASTList(coreForm);

        return new Ident(str);
    } else if (inp instanceof OperatorToken ot) {
        for (val oper : syntax.coreOperators) {
            if (arraysEqual(ot.val, oper.i0)) {
                return new CoreOperatorAST(oper.i1);
            }
        }
        for (val oper : syntax.functionOperators) {
            if (arraysEqual(ot.val, oper.i0)) {
                return new FunctionOperatorAST(oper.i1);
            }
        }
        return new OperatorAST(ot.val);
    } else if (inp instanceof CommentToken ct) {
        return new CommentAST(ct.val);
    } else {
        // should never happen
        return new IntLiteral(-1);
    }
}

}
